src/linker/Linker.Steps/EmbeddedXmlStep.cs

```diff
@@ -37,76 +37,104 @@
 		{
-			var steps_to_add = new Stack<IStep> ();
+			var steps = new List<ResolveFromXmlStep> ();
+			var embeddedXml = GetEmbeddedXml (_assembly);
 
 #if !ILLINK
 			foreach (string name in Assembly.GetExecutingAssembly ().GetManifestResourceNames ()) {
-				if (!name.EndsWith (".xml", StringComparison.OrdinalIgnoreCase) || !ShouldProcessRootDescriptorResource (GetAssemblyName (name)))
+				if (!name.EndsWith (".xml", StringComparison.OrdinalIgnoreCase) || !ShouldProcessRootDescriptorResource (_assembly, name))
 					continue;
 
 				try {
 					Context.LogMessage ($"Processing resource linker descriptor: {name}");
-					steps_to_add.Push (GetResolveStep (name));
+					steps.Add (GetResolveStep (name, markStep));
 				} catch (XmlException ex) {
 					/* This could happen if some broken XML file is included. */
 					Context.LogError ($"Error processing {name}: {ex}", 1003);
 				}
 			}
 #endif
-
-			foreach (var asm in Context.GetAssemblies ()) {
-				if (Annotations.GetAction (asm) == AssemblyAction.Skip)
-					continue;
-
-				var embeddedXml = asm.Modules
-					.SelectMany (mod => mod.Resources)
-					.Where (res => res.ResourceType == ResourceType.Embedded)
-					.Where (res => res.Name.EndsWith (".xml", StringComparison.OrdinalIgnoreCase))
-					.ToArray ();
-
+			if (Annotations.GetAction (_assembly) != AssemblyAction.Skip) {
 				foreach (var rsc in embeddedXml
-									.Where (res => ShouldProcessRootDescriptorResource (res.Name))
-									.Cast<EmbeddedResource> ()) {
+							.Where (res => ShouldProcessRootDescriptorResource (_assembly, res.Name))
+							.Cast<EmbeddedResource> ()) {
 					try {
-						Context.LogMessage ($"Processing embedded linker descriptor {rsc.Name} from {asm.Name}");
-						steps_to_add.Push (GetExternalResolveStep (rsc, asm));
+						Context.LogMessage ($"Processing embedded linker descriptor {rsc.Name} from {_assembly.Name}");
+						steps.Add (GetExternalResolveStep (rsc, _assembly, markStep));
 					} catch (XmlException ex) {
 						/* This could happen if some broken XML file is embedded. */
 						Context.LogError ($"Error processing {rsc.Name}: {ex}", 1003);
 					}
 				}
+			}
 
-				foreach (var rsc in embeddedXml
-									.Where (res => res.Name.Equals ("ILLink.Substitutions.xml", StringComparison.OrdinalIgnoreCase))
-									.Cast<EmbeddedResource> ()) {
-					try {
-						Context.LogMessage ($"Processing embedded substitution descriptor {rsc.Name} from {asm.Name}");
-						steps_to_add.Push (GetExternalSubstitutionStep (rsc, asm));
-					} catch (XmlException ex) {
-						Context.LogError ($"Error processing {rsc.Name}: {ex}", 1003);
-					}
+			foreach (var step in steps) {
+				step.Process (Context);
+			}
+		}
+
+		public void ProcessSubstitutions ()
+		{
+			if (Annotations.GetAction (_assembly) == AssemblyAction.Skip)
+				return;
+
+			var steps = new List<BodySubstituterStep> ();
+			var embeddedXml = GetEmbeddedXml (_assembly);
+
+			foreach (var rsc in embeddedXml
```

Comment:
Why do you need foreach when you are looking for single entry name? Wouldn't be better to change `GetEmbeddedXml` to return single entry?

Reply:
This was existing code that I moved around - I cleaned it up a bit now.